crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -0,0 +1,302 @@
+            total_keys_scanned, total_objects_scanned
+        );
+        total_pruned
+    }
+
+    fn setup_objects_pruning(
+        num_versions_to_retain: u64,
+        pruning_timeperiod: Duration,
+        pruning_initial_delay: Duration,
+        perpetual_db: Arc<AuthorityPerpetualTables>,
+    ) -> Sender<()> {
+        let (sender, mut recv) = tokio::sync::oneshot::channel();
+        if num_versions_to_retain == u64::MAX {
+            info!("Skipping pruning of objects table as we want to retain all versions");
+            return sender;
+        }
+        let mut prune_interval =
+            tokio::time::interval_at(Instant::now() + pruning_initial_delay, pruning_timeperiod);
+        prune_interval.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+        tokio::task::spawn(async move {
+            loop {
+                tokio::select! {
+                    _ = prune_interval.tick() => {
+                        info!("Starting pruning of objects table");
+                        let perpetual_db = perpetual_db.clone();
+                        let num_pruned = Self::prune_objects(num_versions_to_retain, perpetual_db);
+                        info!("Finished pruning with total object versions pruned = {}", num_pruned);
+                    }
+                    _ = &mut recv => break,
+                }
+            }
+        });
+        sender
+    }
+    pub fn new(
+        perpetual_db: Arc<AuthorityPerpetualTables>,
+        pruning_config: &AuthorityStorePruningConfig,
+    ) -> Self {
+        AuthorityStorePruner {
+            _objects_pruner_cancel_handle: Self::setup_objects_pruning(
+                pruning_config.objects_num_latest_versions_to_retain,
+                Duration::from_secs(pruning_config.objects_pruning_period_secs),
+                Duration::from_secs(pruning_config.objects_pruning_initial_delay_secs),
+                perpetual_db,
+            ),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{collections::HashSet, sync::Arc};
+
+    use crate::authority::{
+        authority_store::ObjectKey, authority_store_tables::AuthorityPerpetualTables,
+    };
+    use fs_extra::dir::get_size;
+    use more_asserts as ma;
+    use sui_types::{
+        base_types::{ObjectID, SequenceNumber},
+        object::Object,
+    };
+    use tracing::log::info;
+    use typed_store::Map;
+
+    use super::AuthorityStorePruner;
+
+    async fn test_pruning(
+        perpetual_db: Arc<AuthorityPerpetualTables>,
+        num_versions_per_object: u64,
+        num_object_versions_to_retain: u64,
+        total_unique_object_ids: u32,
+    ) -> Result<u64, anyhow::Error> {
+        // this contains the set of keys that should not have been pruned
+        let mut expected = HashSet::new();
+        let ids = ObjectID::in_range(ObjectID::ZERO, total_unique_object_ids.into())?;
+        for id in ids {
+            for (counter, i) in (0..num_versions_per_object).rev().enumerate() {
```

Comment:
why not just rely on `i`?

Reply:
oh because `counter` is going reverse from `3,2,1,0` and `i` is going `0,1,2,3`. It was just a little simple to do it this way